controllers/storageautoscaler/verify.go

```diff
@@ -0,0 +1,137 @@
+	}
+
+	startOsdSize := storageAutoScaler.Status.LastExpansion.StartOsdSize
+	expectedOsdSize := storageAutoScaler.Status.LastExpansion.ExpectedOsdSize
+	expectedOsdCount := storageAutoScaler.Status.LastExpansion.ExpectedOsdCount
+	expectedStorageCapacity := storageAutoScaler.Status.LastExpansion.ExpectedStorageCapacity
+
+	if startOsdSize.Cmp(expectedOsdSize) != 0 {
+		// vertical scaling
+		err := r.verifyVerticalScaling(ctx, storageAutoScaler, expectedOsdSize)
+		if err != nil {
+			r.Log.Error(err, "failed to verify vertical scaling")
+			return reconcile.Result{RequeueAfter: 1 * time.Minute}, nil
+		}
+	} else {
+		// horizontal scaling
+		err := r.verifyHorizontalScaling(ctx, storageAutoScaler, expectedOsdCount)
+		if err != nil {
+			r.Log.Error(err, "failed to verify horizontal scaling")
+			return reconcile.Result{RequeueAfter: 1 * time.Minute}, nil
+		}
+	}
+
+	// update the status
+	originalStorageAutoScaler := storageAutoScaler.DeepCopy()
+	storageAutoScaler.Status.Phase = "Succeeded"
+	storageAutoScaler.Status.Error = nil
+	storageAutoScaler.Status.LastExpansion.CompletionTime = metav1.Now()
+	storageAutoScaler.Status.LastExpansion.StartOsdSize = expectedOsdSize
+	storageAutoScaler.Status.LastExpansion.StartOsdCount = expectedOsdCount
+	storageAutoScaler.Status.LastExpansion.StartStorageCapacity = expectedStorageCapacity
+
+	err := r.updateStatus(ctx, storageAutoScaler, originalStorageAutoScaler)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	r.Log.Info("scaling verified")
+
+	return reconcile.Result{}, nil
+}
+
+func (r *StorageAutoscalerReconciler) verifyVerticalScaling(ctx context.Context, storageAutoScaler *ocsv1.StorageAutoScaler, expectedOsdSize resource.Quantity) error {
+	osdCountQuery := "(ceph_osd_metadata * on (ceph_daemon, namespace, managedBy) group_right(device_class,hostname) (ceph_osd_stat_bytes))"
+	metrics, err := scrapeMetrics(ctx, r.OperatorNamespace, osdCountQuery, r.Log)
+	if err != nil {
+		r.Log.Error(err, "failed to scrape metrics")
+		return err
+	}
+
+	verified := false
+	for _, osd := range metrics {
+		deviceClass := osd.Metric["device_class"]
+		if string(deviceClass) == storageAutoScaler.Spec.DeviceClass {
+			verified = true
+			if expectedOsdSize.Cmp(resource.MustParse(osd.Value.String())) > 0 {
+				r.Log.Info("osd size not as expected", "expected", expectedOsdSize, "actual", osd.Value)
+				return errors.New("osd size not as expected, expected: " + expectedOsdSize.String() + " actual: " + osd.Value.String())
+			}
+			break
+		}
+	}
+
+	if !verified {
+		r.Log.Info("device class not found in metrics", "device class", storageAutoScaler.Spec.DeviceClass)
+		return errors.New("device class not found in metrics, device class: " + storageAutoScaler.Spec.DeviceClass)
+	}
+
+	return nil
+}
+
+func (r *StorageAutoscalerReconciler) verifyHorizontalScaling(ctx context.Context, storageAutoScaler *ocsv1.StorageAutoScaler, expectedOsdCount uint16) error {
```

Comment:
I think this should have a unit test. I'm a little confused by the logic here, and it would be good to see a unit test to know what the function is expected to be doing

Reply:
I would like to take this as a stretch goal or a follow up if that's okay? As internally it called `scarpeMetrics()` which haven't been unit tested, and doesn't feel like straight forward unit for `scarpeMetric` function

Reply:
ocs-operator maintainers can determine their level of code risk when code isn't unit tested